source/fb_nextgen_mmm_v20.0.func.R

```diff
@@ -741,16 +747,25 @@ f.mmm <- function(...
 
   ## set paralle backend
   cat("\nRunning", iterRS,"random search trails with",lambda.n,"trails lambda cross-validation each on",set_cores,"cores...\n")
-  cl <- makeCluster(set_cores)# makeSOCKcluster(set_cores) #makeCluster(set_cores)
-  registerDoSNOW(cl)
+  #cl <- makeCluster(set_cores)# makeSOCKcluster(set_cores) #makeCluster(set_cores)
+  #registerDoSNOW(cl)
   pb <- txtProgressBar(max = iterRS, style = 3)
   opts <- list(progress = function(n) setTxtProgressBar(pb, n))
 
-  getDoParWorkers()
+  #getDoParWorkers()
   
   ################################################
+  ng <- import("nevergrad")
   #### Start parallel loop
   t0 <- Sys.time()
+
+  # Creating an optimizer.
+  my_tuple <- tuple(length(hyperParams[[1]]))
+  instrumentation <- ng$p$Array(shape=my_tuple)
+  instrumentation$set_bounds(0., 1.)
+  optimizer <-  ng$optimizers$registry["DiscreteOnePlusOne"](instrumentation)  # length(hyperParams[[1]]))
+  # Creating an hyperparameter vector to be used in the next learning.
+
   sysTimeDopar <- system.time({
```

Comment:
It seems he is computing in parallel anyways?

Reply:
I think it does not.
It does not mean that it's necessarily an issue - the speed-up due to switching to Nevergrad is much greater than the parallel speed-up. However, as Nevergrad is fundamentally able to run in parallel (including asynchronously) it would be bettter to make this parallel. I would find that easy in Python but I need to learn a bit of R syntax for doing it.

What we need to do (for 10 cores) if we don't care about asynchronous optimization:
- add an argument "num_workers = 10" in line 753 "optimizer <- ...., budget=iterRS", so that the algorithm knows it must be parallel (some algorithms don't care)
- do 10 times "ask" (sequentially, this is anyway super fast)
- compute in parallel the 10 "mape" corresponding to those 10 nevergrad_hp
- do 10 times "tell" (sequentially, this is anyway super fast)

Reply:
Great Oli thanks so much for your explanation. One more for the backlog! Maybe we can add this once we tested the objective function with more than one variable.